flink-libraries/flink-table/src/main/scala/org/apache/flink/table/calcite/RelTimeIndicatorConverter.scala

```diff
@@ -165,7 +165,30 @@ class RelTimeIndicatorConverter(rexBuilder: RexBuilder) extends RelShuttle {
   override def visit(exchange: LogicalExchange): RelNode =
     throw new TableException("Logical exchange in a stream environment is not supported yet.")
 
-  override def visit(scan: TableScan): RelNode = scan
+  override def visit(scan: TableScan): RelNode = {
+    val upsertStreamTable = scan.getTable.unwrap(classOf[UpsertStreamTable[_]])
+    val relTypes = scan.getRowType.getFieldList.map(_.getType)
+    val getTimeIndicatorIndexes = relTypes.zipWithIndex
+      .filter(e => FlinkTypeFactory.isTimeIndicatorType(e._1))
+      .map(_._2)
+    if (upsertStreamTable != null) {
+      val input = if (getTimeIndicatorIndexes.size > 0) {
+        // materialize time indicator
+        val rewrittenScan = scan.copy(scan.getTraitSet, scan.getInputs)
+        materializerUtils.projectAndMaterializeFields(rewrittenScan, getTimeIndicatorIndexes.toSet)
+      } else {
+        scan
+      }
+
+      LogicalLastRow.create(
```

Comment:
Year, It would be nice if we can. A rule can only transform from a node to a new node with a same rowtype. However, the rowtype of LastRow is different from the scan, i.e, the tuple type has been flattened. I suggest to

Reply:
But you could probably add this node when changing convention. We assume that `TabelScan` rel node represents all kinds of table scans and it always can produce retractions, while for some stage (`FlinkLogical`?) we freeze it to either retractions/upserts.

Regardless of that, `RelTimeIndicatorConverter.scala` is not the right place to create `LogicalLastRow`. You could also add a single pass rule, that is applied only once, that inserts upsert -> retraction conversion.